cpp/include/cudf/fixed_point/fixed_point.hpp

```diff
@@ -0,0 +1,299 @@
+    BASE_10 = 10
+};
+
+namespace detail {
+    // helper function to negate strongly typed scale_type
+    auto negate(scale_type const& scale) -> scale_type {
+        return scale_type{-scale};
+    }
+
+    // perform this operation when constructing with positive scale
+    template <Radix Rad, typename T>
+    constexpr auto right_shift(T const& val, scale_type const& scale) -> double {
+        assert(scale > 0);
+        return val / std::pow(static_cast<int32_t>(Rad), static_cast<int32_t>(scale));
+    }
+
+    // perform this operation when constructing with negative scale
+    template <Radix Rad, typename T>
+    constexpr auto left_shift(T const& val, scale_type const& scale) -> double {
+        assert(scale < 0);
+        return val * std::pow(static_cast<int32_t>(Rad), static_cast<int32_t>(negate(scale)));
+    }
+
+    // convenience generic shift function
+    template <Radix Rad, typename T>
+    constexpr auto shift(T const& val, scale_type const& scale) -> double {
+        if      (scale == 0) return static_cast<double>(val);
+        else if (scale >  0) return right_shift<Rad>(val, scale);
+        else                 return left_shift <Rad>(val, scale);
+    }
+}
+
+template <typename T>
+constexpr inline auto is_supported_representation_type() -> bool {
+  return std::is_same<T, int32_t>::value;
+}
+
+template <typename T>
+constexpr inline auto is_supported_construction_value_type() -> bool {
+  return std::is_integral      <T>::value
+      || std::is_floating_point<T>::value;
+}
+
+// helper struct for constructing fixed_point when value is already shifted
+template <typename Rep,
+          typename std::enable_if_t<is_supported_representation_type<Rep>()>* = nullptr>
+struct scaled_integer{
+    Rep value;
+    scale_type scale;
+    explicit scaled_integer(Rep v, scale_type s) : value(v), scale(s) {}
+};
+
+// Rep = representative type
+template <typename Rep, Radix Rad>
+class fixed_point {
+
+    Rep        _value;
+    scale_type _scale;
+
+public:
+
+    // CONSTRUCTORS
+    template <typename T,
+              typename std::enable_if_t<is_supported_construction_value_type<T>()
+                                     && is_supported_representation_type<Rep>()>* = nullptr>
+    explicit fixed_point(T const& value, scale_type const& scale) :
+        _value{static_cast<Rep>(detail::shift<Rad>(value, scale))},
+        _scale{scale}
+    {
+    }
+
+    explicit fixed_point(scaled_integer<Rep> s) :
+        _value{s.value},
+        _scale{s.scale}
+    {
+    }
+
+    // DEFAULT CONSTRUCTOR
+    fixed_point() :
+        _value{0},
+        _scale{scale_type{0}}
+    {
+    }
+
+    // EXPLICIT CONVERSION OPERATOR
+    template <typename U,
+              typename std::enable_if_t<is_supported_construction_value_type<U>()>* = nullptr>
+    explicit constexpr operator U() const {
+        return detail::shift<Rad>(static_cast<U>(_value), detail::negate(_scale));
+    }
+
+    auto get() const noexcept {
```

Comment:
Is `get` the right name for this? It's not super clear what it means to `get` a `fixed_point`.  Perhaps:

```suggestion
    double to_double() const noexcept {
```

But actually, this just calls the above explicit conversion operator, right? In that case is this function even necessary?

Reply:
The alternatives are:
1. Have the user know to `static_cast` to desired type
2. Have free functions called `to_int` / `to_double`
3. Have member functions called `to_int` / `to_double`
4. Have template member `to<>` but require `fp.template to<int32_t>` at call site

I agree, `get` isn't desirable. I chose option 3 because 1 is not easy to use / easy for the user to know they have to do it, 2 has similar reasons to 1 but also doesn't read as nicely as the member function and 4 is just awful. 

Let me know if you have a different opinion, I don't feel to strongly between 1, 2 and 3.

The change was pushed this morning.